MmSupervisorPkg/Core/PrivilegeMgmt/CallGateTransfer.c

```diff
@@ -195,21 +195,132 @@ CallgateInit (
   UINTN                 CpuIndex;
   EFI_PHYSICAL_ADDRESS  GdtrBaseAddr;
 
-  PrivilegeMgmtFixupAddress ();
-
   for (CpuIndex = 0; CpuIndex < NumberOfCpus; CpuIndex++) {
     if (CpuIndex == mSmmMpSyncData->BspIndex) {
       // BSP call gate will change, patch at runtime
       continue;
     }
 
-    // Patch AP handlers call gate and stack here as they are static after init
+    // Patch AP handlers call gate here, they can still change during later usage
     GdtrBaseAddr = mGdtBuffer + mGdtStepSize * CpuIndex;
-    PatchCallGatePtr ((IA32_IDT_GATE_DESCRIPTOR *)(UINTN)(GdtrBaseAddr + CALL_GATE_OFFSET), (VOID *)ApHandlerReturnPointer);
+    PatchCallGatePtr ((IA32_IDT_GATE_DESCRIPTOR *)(UINTN)(GdtrBaseAddr + CALL_GATE_OFFSET), (VOID *)NULL);
     PatchTssDescriptor (
       (IA32_TSS_DESCRIPTOR *)(UINTN)(GdtrBaseAddr + TSS_SEL_OFFSET),
       (IA32_TASK_STATE_SEGMENT *)(UINTN)(GdtrBaseAddr + TSS_DESC_OFFSET),
       (VOID *)GetThisCpl3Stack (CpuIndex)
       );
   }
 }
+
+/**
+  Invoke MM driver in CPL 3.
+**/
+EFI_STATUS
+EFIAPI
+InvokeDemotedDriverEntryPoint (
+  IN MM_IMAGE_ENTRY_POINT  *EntryPoint,
+  IN EFI_HANDLE            ImageHandle,
+  IN EFI_MM_SYSTEM_TABLE   *MmSystemTable

```

Comment:
Are there cases where this function is invoked and `MmSystemTable` is `NULL`?

Reply:
Yes, the first driver (ideally `Ring3Broker`) loaded will have a NULL pointer for this field. This is because the table initiated in MM core is useless for the user drivers directly due to privilege mismatch.

The user version of system table is registered by the `Ring3Broker` once it is loaded and all the subsequent module entrypoints will be loaded with the user version of `MmSystemTable`.